samcli/cli/cli_config_file.py

```diff
@@ -0,0 +1,185 @@
+    This function also figures out the command name by looking up current click context's parent
+    and constructing the parsed command name that is used in default configuration file.
+    If a given cmd_name is start-api, the parsed name is "local_start_api".
+    provider is called with `config_file`, `config_env_name` and `parsed_cmd_name`.
+
+    :param cmd_name: `sam` command name
+    :param provider: provider to be called for reading configuration file
+    :param ctx: Click context
+    :param config_env_name: config-env within configuration file
+    :return: dictionary of defaults for parameters
+    """
+    config_file = os.path.join(os.getcwd(), DEFAULT_CONFIG_FILE_NAME)
+    config = {}
+    if os.path.isfile(config_file):
+        LOG.debug("Config file location: %s", os.path.abspath(config_file))
+
+        # Find parent of current context
+        _parent = ctx.parent
+        _cmd_names = []
+        # Need to find the total set of commands that current command is part of.
+        if cmd_name != ctx.info_name:
+            _cmd_names = [cmd_name]
+        _cmd_names.append(ctx.info_name)
+        # Go through all parents till a parent of a context exists.
+        while _parent.parent:
+            info_name = _parent.info_name
+            _cmd_names.append(info_name)
+            _parent = _parent.parent
+
+        # construct a parsed name that is of the format: a_b_c_d
+        parsed_cmd_name = "_".join(reversed([cmd.replace("-", "_").replace(" ", "_") for cmd in _cmd_names]))
+
+        config = provider(config_file, config_env_name, parsed_cmd_name)
+
+    return config
+
+
+def configuration_option(*param_decls, **attrs):
+    """
+    Adds configuration file support to a click application.
+
+    This will create an option of type `STRING` expecting the config_env in the
+    configuration file, by default this config_env is `default`. When specified,
+    the requisite portion of the configuration file is considered as the
+    source of truth.
+
+    The default name of the option is `--config-env`.
+
+    This decorator accepts the same arguments as `click.option`.
+    In addition, the following keyword arguments are available:
+    :param cmd_name: The command name. Default: `ctx.info_name`
+    :param config_env_name: The config_env name. This is used to determine which part of the configuration
+        needs to be read.
+    :param provider:         A callable that parses the configuration file and returns a dictionary
+        of the configuration parameters. Will be called as
+        `provider(file_path, config_env, cmd_name)
+    """
+    param_decls = param_decls or ("--config-env",)
+    option_name = param_decls[0]
+
+    def decorator(f):
+
+        attrs.setdefault("is_eager", True)
+        attrs.setdefault("help", "Read config-env from Configuration File.")
+        attrs.setdefault("expose_value", False)
+        # --config-env is hidden and can potentially be opened up in the future.
```

Comment:
can we make sure we don't respect user specififc value?

Reply:
Done.